crates/wasmparser/src/validator/component_types.rs

```diff
@@ -34,56 +35,174 @@ const MAX_FLAT_FUNC_RESULTS: usize = 1;
+    fn new(
+        types: &'a TypeList,
+        desc: &'static str,
+        variance: Variance,
+        offset: usize,
+        expected: &[ValType],
+    ) -> Self {
+        Self {
+            types,
+            expected: expected.iter().copied().collect(),
+            index: 0,
+            desc,
+            variance,
+            offset,
+        }
+    }
+
+    fn reset(&mut self) {
+        self.index = 0;
+    }
+
+    fn peek(&self) -> Option<ValType> {
+        self.expected.get(self.index).copied()
+    }
+
+    fn match_ty(&mut self, actual: ValType) -> LowerResult {
+        let expected = match self.peek() {
+            Some(e) => e,
+            // Arity mismatch, but ignore it for now. The arity check at the
+            // end of `ComponentFuncType::lower` will catch this and can
+            // give a more descriptive error than we can here.
+            None => return Ok(()),
+        };
+
+        let (expected, actual) = match self.variance {
+            Variance::Covariant => (expected, actual),
+            Variance::Contravariant => (actual, expected),
+        };
+
+        if Matches::matches(self.types, actual, expected) {
+            self.index += 1;
+            return Ok(());
+        }
+
+        let desc = self.desc;
+        let index = self.index;
+        Err(LowerError::TypeMismatch(format_err!(
+            self.offset,
+            "type mismatch when checking `core-type` canonical option in {desc}[{index}]:\n\
+             expected: {expected}\n\
+             actual:   {actual}",
+        )))
+    }
+}
+
 /// A simple alloc-free list of types used for calculating lowered function signatures.
-pub(crate) struct LoweredTypes {
+struct LoweredTypes<'a> {
+    expected: Option<ExpectedLoweredTypes<'a>>,
     types: [ValType; MAX_LOWERED_TYPES],
     len: usize,
     max: usize,
 }
 
-impl LoweredTypes {
-    fn new(max: usize) -> Self {
+impl<'a> LoweredTypes<'a> {
+    fn new(expected: Option<ExpectedLoweredTypes<'a>>, max: usize) -> Self {
         assert!(max <= MAX_LOWERED_TYPES);
         Self {
+            expected,
             types: [ValType::I32; MAX_LOWERED_TYPES],
             len: 0,
             max,
         }
     }
 
-    fn len(&self) -> usize {
-        self.len
+    fn clone_without_expected_types(&self) -> Self {
+        Self {
+            expected: None,
+            types: self.types,
+            len: self.len,
+            max: self.max,
+        }
     }
 
     fn maxed(&self) -> bool {
         self.len == self.max
     }
 
-    fn get_mut(&mut self, index: usize) -> Option<&mut ValType> {
-        if index < self.len {
-            Some(&mut self.types[index])
-        } else {
-            None
-        }
-    }
-
+    /// Like `try_push` but asserts that this list has capacity for the type.
     #[track_caller]
-    fn assert_push(&mut self, ty: ValType) {
-        assert!(self.try_push(ty));
+    fn assert_push(&mut self, ty: ValType) -> Result<()> {
```

Comment:
I think we should probably rename this to something else since it's returning a `Result`. 

How about renaming `try_push` to `push`, removing `assert_push`, and using test coverage to test the `assert_*` cases? Everything already returns a `Result` anyway so using `?` liberally seems fine here.

Reply:
`try_push` and `assert_push` return two different kinds of results tho, which actually works out pretty nice right now. `try_push` returns `LowerResult`s and notably that means it can return `OutOfCapacity` errors, while `assert_push` returns a `Result<(), BinaryReaderError>` and can never fail with out-of-capacity errors. This is useful for stratifying the try-and-lower-into-a-max-size-arg-list use cases versus the infallible indirect-through-memory use cases.

Reply:
Don't love the names either tho.